### DISCUSSION TITLE
scala_textへの感想

### DISCUSSION THREAD

**Discussion by kmizu** [scala-text/scala_text#235] [General]
- このdiscussionは本テキストへの感想を書くためのものです。
  - 返信はあまり期待しないでください（返信するかもしれませんが）
- 誤字・脱字の報告、誤りの指摘・修正は別途issueを立てるかpull requestを送ってください
- 感想を書いてもらえると、本テキストのコミッタのモチベーションが上がるかもしれません

**Comment by cm-wada-yusuke** [comment]
クラスメソッド株式会社の社内勉強会にてScala研修テキスト使わせていただき、週一度のペース、約半年で読み終えました。まずは、とてもわかりやすく、かつScalaの面白さが存分に伝わる資料を作っていただき、そしてそれを公開していただいたことに感謝致します。以下、感想です。

* 章立ての構成が良く、「Scalaって何？」という参加者も問題なく順を追って学習することができた
* Scalaの魅力ともいえるコレクションとパターンマッチはやはり厚めに記載されており、力のいれようが参加者にも伝わった
* 結果、「文脈を型で表現し、型に対してパターンマッチをはじめとした操作を適用する」という考えが魅力に感じたようで、swiftやjavaにも欲しいという声が多数
* 成果として、Scalaプログラム（Play製のサーバサイドアプリ）のレビューが可能になる程度になり、開発者としても助かっている
* ただ、「型クラス」については「？」という感じでした。なので「紹介」にとどめておいたのは良かったかと思います。やはりここは言語の機能というより数学的な概念から把握したほうが近道になりうるレベルの話なので仕方ないといえば仕方ないかも
* 「Futureで非同期処理の文脈を扱える」「さらにflatMapで未来の戻り値に対する処理を書ける」「さらにfor-comprehensionでフラットに書ける」あたりが一番盛り上がりました

**Comment by kmizu** [comment]
wadaさん、まずは丁寧なご感想をありがとうございました。元々は、社内研修用のScalaテキストを
ブラッシュアップして公開したものではありますが、自由に使ってもらって感想をもっと欲しいと
思っており（あくまで私がそうであり、他の共著者がどうかはわかりませんが（汗））、とてもうれしいです。

* 一応、本テキストにおける私の役割としては、だいたい全体の1/2~2/3程度執筆、という感じでしょうか。

* 章立てについては、何度か初期の頃に何度か再構成したような記憶があるので、その点を評価していただけるのはとてもありがたいです。

* コレクションに関しては、Scalaの標準ライブラリで最も手厚い部分であるため、ここはかなり力を入れて書いた部分です。パターンマッチは割と普通に書いた気がするのですが、好評なようで良かったです。

* 成果として、Scala Play 2のサーバサイドアプリのレビューが可能になった、というのは想定していなかった部分ですが、お役に立てたようなら何よりです。

* Futureに関しては私はあまり手を出せていなかった部分ですが、「非同期処理の文脈を扱える」「for-comprehensionでフラットに書ける」などの部分が伝わっていたようなのは良かったです。

* 「型クラス」に関しては、正直ご期待に添えず申し訳ありませんでした。ただ、「型クラス」の紹介については、あくまでそういうものがあるよ、という話で、数学的な概念は関係なく便利な道具として使えるよ、というのが本来伝えたかったものではあります（その片鱗がimplicitの章にあります）。ともあれ、感想を全面的に反映するわけにはいきませんが、今後の参考にさせていただきます。

* なお、私が公開部分の多くを執筆したこともあり、私が返答していますが、他に複数名の協力があり、本テキストが成り立っています（名前を出していいのかわからないのでここでは伏せますが）。ご感想、どうもありがとうございました。

**Comment by bleis-tift** [comment]
本文というより、ソース(markdown)の話なのですが、改行の仕方に統一感がないのが気になりました。
INTRODUCTION.mdでは文ですら改行されていませんが、sbt-install.mdでは文の途中でも改行されている、といった具合です。
前者は前者で差分表示の際につらい感じになるのですが、後者は出来上がるgitbookでみると文の途中に空白が入るので見栄えが悪いです。
最低限、句点では改行し、一文が長くなった場合は読点の位置でのみ改行する等のルールを設けた方がいいと思います。

また、明確な誤りではないものはこちらに書くようになっているように読めたのでここに書きますが、[sbt-install.md](https://github.com/dwango/scala_text/blob/739514dd20a9e4fb7293f125a667b1b4dc441ced/src/sbt-install.md)に「Windowsでの環境変数編集ツールとしては、 Rapid Environment Editorが非常に便利です。」とありました。
Windows10からはPATHを編集する程度であれば標準のもので十分に思えますので、この記述は不要ではないかと思うのですがいかがでしょうか。

**Comment by bleis-tift** [comment]
[control-syntax.md](https://github.com/dwango/scala_text/blob/daf85006396150fffaf43d575a1ae34037c379e5/src/control-syntax.md)のwhile式の説明で、

> なお、`break`文や`continue`文に相当するものはありません。

とありますが、breakやcontinueがしたくなった場合にどうするかということについて書いてあった方が親切だと思います。

**Comment by bleis-tift** [comment]
構文の説明に使っている独自形式に説明も統一感もないのは問題だと思います。

**Comment by kmizu** [comment]
> とありますが、breakやcontinueがしたくなった場合にどうするかということについて
> 書いてあった方が親切だと思います。

その直後でフォローするのは難しい（メソッド内メソッドを定義するとか、高階関数を使うとかいう話になるので）ですが、フォローできるかは検討してみます。

**Comment by kmizu** [comment]
> Windows10からはPATHを編集する程度であれば標準のもので十分に思えますので、この記述は
> 不要ではないかと思うのですがいかがでしょうか。

個人的には、Windows 10のでもまだめんどくさい印象があるのですが、標準でないツールを
あんまり入れさせるのはどうかという話はあるかなという気がしています。とりあえず、現状で
変更する程ではないかなと思っています。

**Comment by kmizu** [comment]
@bleis-tift さん

https://github.com/dwango/scala_text/issues/235#issuecomment-335213039

>  構文の説明に使っている独自形式に説明も統一感もないのは問題だと思います。

かなり対応が遅くなって申し訳ありませんが、新たに「記法」の章を設けて、構文の説明に使っている構文（メタ構文）について説明を追加し、それに従うように説明を書き直してみました。どうでしょうか？

**Comment by xuwei-k** [comment]
https://github.com/scala-text/scala_text/discussions

新機能のdiscussionsを試しに有効にしてみたけど、このissueはdiscussionsにconvertするか？ 🤔

**Comment by xuwei-k** [comment]
convertは自動で出来るっぽいです。(右側の方)
おそらく、そういうことを意図して、issueやpull reqと、インクリメンタルされるidが共用みたいですし

![スクリーンショット 2020-12-11 1 10 51](https://user-images.githubusercontent.com/389787/101797759-baf3ea00-3b4d-11eb-80a4-2a302d339d2b.png)

**Comment by kmizu** [comment]
@xuwei-k 問題なさげなので、それ見てconvertしてみました（既にconvertされている）。
